pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,753 @@
+	ofClient                    openflow.Client
+	interfaceStore              interfacestore.InterfaceStore
+	nodeConfig                  *config.NodeConfig
+	queue                       workqueue.RateLimitingInterface
+	runningPacketSamplingsMutex sync.RWMutex
+	runningPacketSamplings      map[uint8]*packetSamplingState
+	enableAntreaProxy           bool
+	sftpUploader                ftp.UpLoader
+}
+
+func NewPacketSamplingController(
+	kubeClient clientset.Interface,
+	crdClient clientsetversioned.Interface,
+	serviceInformer coreinformers.ServiceInformer,
+	endpointInformer coreinformers.EndpointsInformer,
+	packetSamplingInformer crdinformers.PacketSamplingInformer,
+	client openflow.Client,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	enableAntreaProxy bool,
+) *Controller {
+	c := &Controller{
+		kubeClient:             kubeClient,
+		crdClient:              crdClient,
+		packetSamplingInformer: packetSamplingInformer,
+		packetSamplingLister:   packetSamplingInformer.Lister(),
+		packetSamplingSynced:   packetSamplingInformer.Informer().HasSynced,
+		ofClient:               client,
+		interfaceStore:         interfaceStore,
+		nodeConfig:             nodeConfig,
+		queue: workqueue.NewRateLimitingQueueWithConfig(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay),
+			workqueue.RateLimitingQueueConfig{Name: "packetsampling"}),
+		runningPacketSamplings: make(map[uint8]*packetSamplingState),
+		sftpUploader:           &ftp.SftpUploader{},
+		enableAntreaProxy:      enableAntreaProxy,
+	}
+
+	packetSamplingInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addPacketSampling,
+		UpdateFunc: c.updatePacketSampling,
+		DeleteFunc: c.deletePacketSampling,
+	}, resyncPeriod)
+
+	c.ofClient.RegisterPacketInHandler(uint8(openflow.PacketInCategoryPS), c)
+
+	if c.enableAntreaProxy {
+		c.serviceLister = serviceInformer.Lister()
+		c.serviceListerSynced = serviceInformer.Informer().HasSynced
+		c.endpointLister = endpointInformer.Lister()
+		c.endpointSynced = endpointInformer.Informer().HasSynced
+	}
+	return c
+
+}
+
+func (c *Controller) enqueuePacketSampling(ps *crdv1alpha1.PacketSampling) {
+	c.queue.Add(ps.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the PacketSampling events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting packetsampling controller.", "name", controllerName)
+	defer klog.InfoS("Shutting down packetsampling controller.", "name", controllerName)
+
+	cacheSynced := []cache.InformerSynced{c.packetSamplingSynced}
+	if c.enableAntreaProxy {
+		cacheSynced = append(cacheSynced, c.serviceListerSynced, c.endpointSynced)
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSynced...) {
+		return
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
```

Comment:
`MkdirAll` will not fail if it's already exist. it works like `mkdir -p`. I feel maybe force delete before mkdir should work, to cleanup the possible stale files. Because they are `cache` files,  then end up in the ftp server.